client/network/sync/src/engine.rs

```diff
@@ -0,0 +1,929 @@
+						);
+					}
+				},
+				ToServiceCommand::AnnounceBlock(hash, data) => self.announce_block(hash, data),
+				ToServiceCommand::NewBestBlockImported(hash, number) =>
+					self.new_best_block_imported(hash, number),
+				ToServiceCommand::Status(tx) =>
+					if let Err(_) = tx.send(self.chain_sync.status()) {
+						log::warn!(target: "sync", "Failed to respond to `Status` query");
+					},
+				ToServiceCommand::NumActivePeers(tx) => {
+					if let Err(_) = tx.send(self.chain_sync.num_active_peers()) {
+						log::warn!(target: "sync", "response channel closed for `NumActivePeers`");
+					}
+				},
+				ToServiceCommand::SyncState(tx) => {
+					if let Err(_) = tx.send(self.chain_sync.status()) {
+						log::warn!(target: "sync", "response channel closed for `SyncState`");
+					}
+				},
+				ToServiceCommand::BestSeenBlock(tx) => {
+					if let Err(_) = tx.send(self.chain_sync.status().best_seen_block) {
+						log::warn!(target: "sync", "response channel closed for `BestSeenBlock`");
+					}
+				},
+				ToServiceCommand::NumSyncPeers(tx) => {
+					if let Err(_) = tx.send(self.chain_sync.status().num_peers) {
+						log::warn!(target: "sync", "response channel closed for `NumSyncPeers`");
+					}
+				},
+				ToServiceCommand::NumQueuedBlocks(tx) => {
+					if let Err(_) = tx.send(self.chain_sync.status().queued_blocks) {
+						log::warn!(target: "sync", "response channel closed for `NumQueuedBlocks`");
+					}
+				},
+				ToServiceCommand::NumDownloadedBlocks(tx) => {
+					if let Err(_) = tx.send(self.chain_sync.num_downloaded_blocks()) {
+						log::warn!(target: "sync", "response channel closed for `NumDownloadedBlocks`");
+					}
+				},
+				ToServiceCommand::NumSyncRequests(tx) => {
+					if let Err(_) = tx.send(self.chain_sync.num_sync_requests()) {
+						log::warn!(target: "sync", "response channel closed for `NumSyncRequests`");
+					}
+				},
+				ToServiceCommand::PeersInfo(tx) => {
+					let peers_info =
+						self.peers.iter().map(|(id, peer)| (*id, peer.info.clone())).collect();
+					if let Err(_) = tx.send(peers_info) {
+						log::warn!(target: "sync", "response channel closed for `PeersInfo`");
+					}
+				},
+				ToServiceCommand::OnBlockFinalized(hash, header) =>
+					self.chain_sync.on_block_finalized(&hash, *header.number()),
+			}
+		}
+
+		while let Poll::Ready(result) = self.chain_sync.poll(cx) {
+			self.process_block_announce_validation_result(result);
+		}
+
+		Poll::Pending
+	}
+
+	/// Called by peer when it is disconnecting.
+	///
+	/// Returns a result if the handshake of this peer was indeed accepted.
+	pub fn on_sync_peer_disconnected(&mut self, peer: PeerId) -> Result<(), ()> {
+		if self.important_peers.contains(&peer) {
+			log::warn!(target: "sync", "Reserved peer {} disconnected", peer);
+		} else {
+			log::debug!(target: "sync", "{} disconnected", peer);
+		}
+
+		if let Some(_peer_data) = self.peers.remove(&peer) {
```

Comment:
```suggestion
		if self.peers.remove(&peer).is_some() {
```

But yeah, we should remove the duplicated storing of the peers.